File: tests/streaming/markdownHarness.mjs
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// `B882`. This file was the fourth copy of the `markdown.js` import-rewriting
// trick and the one nobody noticed was broken: `B872` added a fifth import to
// `markdown.js`, the three Python copies were each pasted with a fifth inline,
// and this one — under node's own test runner rather than pytest — was left
// raising `ERR_UNSUPPORTED_RESOLVE_REQUEST: Failed to resolve module specifier
// "./markdown/mermaidTheme.js" from "data:text/javascript;base64,…"`, which
// failed the whole streaming suite at import. Its own comments had written
// that sentence twice already, under `B250` and `P5-06`.
//
// The loader is now `tests/helpers/markdownHarness.mjs`, which resolves
// `markdown.js`'s imports by reading them instead of listing them, so a sixth
// import needs no edit anywhere. `loadMarkdown` is re-exported here under the
// name the streaming tests have always imported — a door, not a deletion.
//
// `normalizeRender` stays, because it belongs to these tests and to nothing
// else: it is about comparing two renders of one document, not about loading
// the renderer.
export { loadMarkdown, importMarkdown, installMarkdownDom } from '../helpers/markdownHarness.mjs';

// Canonicalize rendered HTML so two renders that produce the SAME DOM compare
// equal. Collapses only newline-bearing whitespace BETWEEN tags (`>\n\n<` ->
// `><`): it is insignificant in rendered HTML, and incremental finalization
// legitimately emits `\n\n` between two blocks where a single full render emits
// `\n`. Code whitespace is safe because code is HTML-escaped, so significant
// newlines live inside <code> as text (never between a `>` and a `<`). Inline
// single spaces between tags are left alone. Structural differences (two <ul> vs
// one, <ol> vs <ul>) survive normalization and still fail, as they must.
// Mermaid ids embed Date.now(), so they are normalized too.
export function normalizeRender(html) {
  return String(html)
    .replace(/>\s*\n\s*</g, '><')
    .trim()
    .replace(/(mermaid|thinking)-\d+-\d+/g, '$1-X');
}

File: tests/helpers/markdown_harness.py
# SPDX-License-Identifier: AGPL-3.0-or-later
r"""`B882` — the Python spelling of the one `markdown.js` loader.

The loader itself is `tests/helpers/markdownHarness.mjs`, and it is JavaScript
because what it does is JavaScript: read `static/js/markdown.js`, replace its
relative imports with the modules they name, and `import()` the result from a
`data:` URL. A Python transcription of that would be the row again in a second
language.

**The part that made this a row rather than a refactor** is that the helper has
to be reachable from two kinds of caller. `tests/streaming/*.test.mjs` runs
under node's own test runner and simply `import`s it. The three Python tests
build a script and run it with `node --input-type=module -e`, where a *relative*
specifier has no base to resolve against — which is the same
`ERR_UNSUPPORTED_RESOLVE_REQUEST` the four copies existed to dodge. The answer
is an absolute `file://` URL, which resolves from anywhere, computed here once:

    from tests.helpers.markdown_harness import harness_import

    script = harness_import("importMarkdown") + '''
        …the caller's own DOM stub…
        const mod = await importMarkdown();
    '''

One implementation, two spellings of the same `import`. Adding an import to
`static/js/markdown.js` now requires no edit to any test file;
`tests/test_one_markdown_harness.py` proves it by adding one.
"""
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]

#: The one loader. An absolute `file://` URL because `node --input-type=module
#: -e` has no base URL to resolve a relative specifier against.
HARNESS_MJS = ROOT / "tests" / "helpers" / "markdownHarness.mjs"
HARNESS_URL = HARNESS_MJS.as_uri()


def harness_import(*names: str) -> str:
    """An `import { … } from '<file url>';` line for a node `-e` script.

    Names available: `importMarkdown`, `installMarkdownDom`, `loadMarkdown`,
    `bundle`, `bundleUrl`, `stripExports`, `importStatements`, `REPO`,
    `MARKDOWN_JS`.
    """
    assert names, "name at least one export"
    return "import { %s } from '%s';\n" % (", ".join(names), HARNESS_URL)
